### ISSUE TITLE
Assertion failure with UPDATE and ORCA

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#3611]
This test case trips an assertion:
```
drop table if exists dml_heap_r,dml_heap_p;

CREATE TABLE dml_heap_r (col1 int,a int , b int default -1, c text) with oids DISTRIBUTED BY (a);

CREATE TABLE dml_heap_p (col1 int, a numeric, b decimal) DISTRIBUTED BY (a,b);
INSERT INTO dml_heap_p(a,b) SELECT id as a, id as b FROM (SELECT * FROM generate_series(1,2) as id) AS x;
 INSERT INTO dml_heap_p(a,b) VALUES(NULL,NULL);
INSERT INTO dml_heap_r(a,b,c) VALUES(generate_series(1,2),generate_series(1,2),'r');
INSERT INTO dml_heap_r(a,b,c) VALUES(NULL,NULL,NULL);

set optimizer=on;

UPDATE dml_heap_r SET a = dml_heap_r.a FROM dml_heap_p WHERE dml_heap_r.b = dml_heap_p.a;
```

```
DETAIL:  FailedAssertion("!(action == DML_INSERT || action == DML_DELETE)", File: "nodeDML.c", Line: 71)
ERROR:  could not connect to segment: initialization of segworker group failed (cdbgang.c:236)
```

This explains the test failure on the `QP_optimizer-function` test on the gpdb_master concourse pipeline (https://gpdb.data.pivotal.ci/teams/main/pipelines/gpdb_master/jobs/QP_optimizer-functional/builds/28). It started to fail after commit ec3370b019, which removed the Assert nodes from the plan of that query. The table in the QP_optimizer-functional contains a NOT NULL constraint, which used to cause an Assert to be added to the plan. The above test case is a slightly modified version of the failing test in QP_optimizer_functional, without the NOT NULL constraint, which fails even on 5X_stable.

**Comment by vraghavan78** [comment]
@hlinnaka did u investigate why this failure is happening?

**Comment by dhanashreek89** [comment]
@vraghavan78 and I looked into this; and the issue exists even before commit ec3370b. The issue happens only when the target table is created with `with oids`. Following version of the repro query works:
```sql
CREATE TABLE dml_heap_t (col1 int,a int , b int default -1, c text) DISTRIBUTED BY (a);

CREATE TABLE dml_heap_p (col1 int, a numeric, b decimal) DISTRIBUTED BY (a,b);
INSERT INTO dml_heap_p(a,b) SELECT id as a, id as b FROM (SELECT * FROM generate_series(1,2) as id) AS x;
 INSERT INTO dml_heap_p(a,b) VALUES(NULL,NULL);
INSERT INTO dml_heap_t(a,b,c) VALUES(generate_series(1,2),generate_series(1,2),'r');
INSERT INTO dml_heap_t(a,b,c) VALUES(NULL,NULL,NULL);

set optimizer=on;

UPDATE dml_heap_t SET a = dml_heap_t.a FROM dml_heap_p WHERE dml_heap_t.b = dml_heap_p.a;
```

**Comment by hlinnaka** [comment]
On 10/26/2017 01:47 AM, Venkatesh Raghavan wrote:
> @hlinnaka did u investigate why this failure is happening?

No.

- Heikki

**Comment by jchampio** [comment]
I think [this merge-branch PR](https://github.com/greenplum-db/gpdb-postgres-merge/pull/5) may have a potential fix (if I'm understanding it correctly).

**Comment by d** [comment]
Closed by 9f3d490 as part of #3656
